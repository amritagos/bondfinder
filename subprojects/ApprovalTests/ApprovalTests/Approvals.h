#pragma once
#include <string>
#include <functional>
#include <exception>
#include <utility>

#include "ApprovalTests/namers/FileNameSanitizerDisposer.h"
#include "ApprovalTests/reporters/FrontLoadedReporterDisposer.h"
#include "ApprovalTests/core/FileApprover.h"
#include "ApprovalTests/integrations/fmt/FmtToString.h"
#include "ApprovalTests/reporters/DefaultReporterDisposer.h"
#include "ApprovalTests/core/Reporter.h"
#include "ApprovalTests/writers/ExistingFile.h"
#include "ApprovalTests/namers/SubdirectoryDisposer.h"
#include "ApprovalTests/namers/DefaultNamerDisposer.h"
#include "ApprovalTests/core/Options.h"
#include "ApprovalTests/utilities/StringMaker.h"
#include "ApprovalTests/ApprovalUtils.h"

namespace ApprovalTests
{

    // TCompileTimeOptions must have a type ToStringConverter, which must have a method toString()
    template <typename TCompileTimeOptions> class TApprovals
    {
    private:
        TApprovals() = default;

        ~TApprovals() = default;

    public:
        static std::shared_ptr<ApprovalNamer> getDefaultNamer()
        {
            return DefaultNamerFactory::getDefaultNamer()();
        }

        template <typename T>
        using IsNotDerivedFromWriter =
            typename std::enable_if<!std::is_base_of<ApprovalWriter, T>::value,
                                    int>::type;

        /**@name Verifying single objects

         See \userguide{TestingSingleObjects,Testing Single Objects}
         */
        ///@{
        static void verify(const std::string& contents,
                           const Options& options = Options())
        {
            StringWriter writer(options.scrub(contents),
                                options.fileOptions().getFileExtension());
            FileApprover::verify(*options.getNamer(), writer, options.getReporter());
        }

        template <typename T, typename = IsNotDerivedFromWriter<T>>
        static void verify(const T& contents, const Options& options = Options())
        {
            verify(TCompileTimeOptions::ToStringConverter::toString(contents), options);
        }

        template <typename T,
                  typename Function,
                  typename = Detail::EnableIfNotOptionsOrReporter<Function>>
        static void
        verify(const T& contents, Function converter, const Options& options = Options())
        {
            std::stringstream s;
            converter(contents, s);
            verify(s.str(), options);
        }

        /// Note that this overload ignores any scrubber in options
        static void verify(const ApprovalWriter& writer,
                           const Options& options = Options())
        {
            FileApprover::verify(*options.getNamer(), writer, options.getReporter());
        }
        ///@}

        /**@name Verifying containers of objects - supplying an iterator range

         See \userguide{TestingContainers,Testing Containers}
         */
        ///@{
        template <typename Iterator>
        static void
        verifyAll(const std::string& header,
                  const Iterator& start,
                  const Iterator& finish,
                  std::function<void(decltype(*start), std::ostream&)> converter,
                  const Options& options = Options())
        {
            std::stringstream s;
            ApprovalUtils::writeHeader(s, header);
            for (auto it = start; it != finish; ++it)
            {
                converter(*it, s);
                s << '\n';
            }
            verify(s.str(), options);
        }
        ///@}

        /**@name Verifying containers of objects - supplying a container

         See \userguide{TestingContainers,Testing Containers}
         */
        ///@{
        template <typename Container>
        static void verifyAll(
            const std::string& header,
            const Container& list,
            std::function<void(typename Container::value_type, std::ostream&)> converter,
            const Options& options = Options())
        {
            verifyAll<typename Container::const_iterator>(
                header, list.begin(), list.end(), converter, options);
        }

        template <typename Container>
        static void verifyAll(const std::string& header,
                              const Container& list,
                              const Options& options = Options())
        {
            int i = 0;
            verifyAll<Container>(
                header,
                list,
                [&](typename Container::value_type e, std::ostream& s) {
                    s << "[" << i++
                      << "] = " << TCompileTimeOptions::ToStringConverter::toString(e);
                },
                options);
        }

        template <typename Container>
        static void verifyAll(const Container& list, const Options& options = Options())
        {
            verifyAll<Container>("", list, options);
        }
        ///@}

        /**@name Verifying containers of objects - supplying an initializer list

         See \userguide{TestingContainers,Testing Containers}
         */
        ///@{
        template <typename T>
        static void
        verifyAll(const std::string& header,
                  const std::initializer_list<T>& list,
                  std::function<void(typename std::initializer_list<T>::value_type,
                                     std::ostream&)> converter,
                  const Options& options = Options())
        {
            verifyAll<std::initializer_list<T>>(header, list, converter, options);
        }

        template <typename T>
        static void verifyAll(const std::string& header,
                              const std::initializer_list<T>& list,
                              const Options& options = Options())
        {
            verifyAll<std::initializer_list<T>>(header, list, options);
        }

        template <typename T>
        static void verifyAll(const std::initializer_list<T>& list,
                              const Options& options = Options())
        {
            verifyAll<std::initializer_list<T>>("", list, options);
        }
        ///@}

        /**@name Other verify methods
         */
        ///@{

        /*! \brief Verify the text of an exception

            See \userguide{TestingExceptions,testing-exception-messages,Testing exception messages}
         */
        static void
        verifyExceptionMessage(const std::function<void(void)>& functionThatThrows,
                               const Options& options = Options())
        {
            std::string message = "*** no exception thrown ***";
            try
            {
                functionThatThrows();
            }
            catch (const std::exception& e)
            {
                message = e.what();
            }
            verify(message, options);
        }

        /// Verify an existing file, that has already been written out
        static void verifyExistingFile(const std::string& filePath,
                                       const Options& options = Options())
        {
            ExistingFile writer(filePath, options);
            FileApprover::verify(writer.getNamer(), writer, options.getReporter());
        }
        ///@}

        /**@name Customising Approval Tests

         These static methods customise various aspects
         of Approval Tests behaviour.
         */
        ///@{

        /// See \userguide{Configuration,using-sub-directories-for-approved-files,Using sub-directories for approved files}
        static SubdirectoryDisposer
        useApprovalsSubdirectory(const std::string& subdirectory = "approval_tests")
        {
            return SubdirectoryDisposer(subdirectory);
        }

        /// See \userguide{Reporters,registering-a-default-reporter,Registering a default reporter}
        static DefaultReporterDisposer
        useAsDefaultReporter(const std::shared_ptr<Reporter>& reporter)
        {
            return DefaultReporterDisposer(reporter);
        }

        /// See \userguide{Reporters,front-loaded-reporters,Front Loaded Reporters}
        static FrontLoadedReporterDisposer
        useAsFrontLoadedReporter(const std::shared_ptr<Reporter>& reporter)
        {
            return FrontLoadedReporterDisposer(reporter);
        }

        /// See \userguide{Namers,registering-a-custom-namer,Registering a Custom Namer}
        static DefaultNamerDisposer useAsDefaultNamer(NamerCreator namerCreator)
        {
            return DefaultNamerDisposer(std::move(namerCreator));
        }

        /// See \userguide{Namers,converting-test-names-to-valid-filenames,Converting Test Names to Valid FileNames}
        static FileNameSanitizerDisposer useFileNameSanitizer(FileNameSanitizer sanitizer)
        {
            return FileNameSanitizerDisposer(sanitizer);
        }
        ///@}
    };

#ifndef APPROVAL_TESTS_DEFAULT_STREAM_CONVERTER
// begin-snippet: customising_to_string_default_converter
#define APPROVAL_TESTS_DEFAULT_STREAM_CONVERTER StringMaker
// end-snippet
#endif

    // Warning: Do not use CompileTimeOptions directly.
    // This interface is subject to change, as future
    // compile-time options are added.
    template <typename TToString> struct CompileTimeOptions
    {
        using ToStringConverter = TToString;
        // more template types may be added to CompileTimeOptions in future, if we add
        // more flexibility that requires compile-time configuration.
    };

    // Template parameter TToString must have a method toString()
    // This interface will not change, as future compile-time options are added.
    template <typename TToString>
    struct ToStringCompileTimeOptions : CompileTimeOptions<TToString>
    {
    };

    using Approvals =
        TApprovals<ToStringCompileTimeOptions<APPROVAL_TESTS_DEFAULT_STREAM_CONVERTER>>;
}
