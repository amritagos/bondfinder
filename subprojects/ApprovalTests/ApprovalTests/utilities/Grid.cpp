#include "Grid.h"
namespace ApprovalTests
{
    std::string Grid::print(int width,
                            int height,
                            std::function<void(int, int, std::ostream&)> printCell)
    {
        std::stringstream s;
        for (int y = 0; y < height; ++y)
        {
            for (int x = 0; x < width; ++x)
            {
                printCell(x, y, s);
            }
            s << '\n';
        }
        return s.str();
    }
    std::string Grid::print(int width, int height, std::string text)
    {
        return print(
            width, height, [&](int /*x*/, int /*y*/, std::ostream& os) { os << text; });
    }
}
