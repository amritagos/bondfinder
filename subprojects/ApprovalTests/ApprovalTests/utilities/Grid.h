#pragma once
#include <sstream>
#include <functional>

namespace ApprovalTests
{
    class Grid
    {
    public:
        static std::string print(int width,
                                 int height,
                                 std::function<void(int, int, std::ostream&)> printCell);
        static std::string print(int width, int height, std::string text);
    };
}
