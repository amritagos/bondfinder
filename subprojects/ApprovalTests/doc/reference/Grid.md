<a id="top"></a>

# Grid

<!-- toc -->
## Contents

  * [What it is](#what-it-is)
  * [Where is this useful?](#where-is-this-useful)
  * [Usage](#usage)
    * [Conventional Usage](#conventional-usage)
    * [Simple Starting Point](#simple-starting-point)<!-- endToc -->

## What it is

A utility for creating 2D text in a grid format.
Grid abstracts away the printing of the grid, so you can focus on how to represent the contents of each cell.

## Where is this useful?

1. When you want to visualize the contents of 2 dimensional data, (usually for verification purposes).
1. When trying to print/verify Board games, Puzzles & 2D Video games that are based on some kind of grid

## Usage 

### Conventional Usage

Grid is passed a width, height, and a function pointer or lambda to represent the contents of each cell. 

For example:

<!-- snippet: grid_lambda -->
<a id='snippet-grid_lambda'></a>
```cpp
auto grid = ApprovalTests::Grid::print(
    4, 3, [](int x, int y, auto& os) { os << "(" << x << "," << y << ") "; });
```
<sup><a href='/tests/DocTest_Tests/utilities/GridTests.cpp#L21-L24' title='Snippet source file'>snippet source</a> | <a href='#snippet-grid_lambda' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

which produces:

<!-- snippet: GridTests.Print_Coordinates.approved.txt -->
<a id='snippet-GridTests.Print_Coordinates.approved.txt'></a>
```txt
(0,0) (1,0) (2,0) (3,0) 
(0,1) (1,1) (2,1) (3,1) 
(0,2) (1,2) (2,2) (3,2)
```
<sup><a href='/tests/DocTest_Tests/utilities/approval_tests/GridTests.Print_Coordinates.approved.txt#L1-L4' title='Snippet source file'>snippet source</a> | <a href='#snippet-GridTests.Print_Coordinates.approved.txt' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

Notice how `Grid::print()` removes the need for looping over `x` & `y`.

### Simple Starting Point

When developing with [Fake it till you make it](/doc/Glossary.md#fake-it-till-you-make-it) Grid can be used to start by printing a simple grid of repeating text, such as:

<!-- snippet: grid_simple -->
<a id='snippet-grid_simple'></a>
```cpp
auto grid = ApprovalTests::Grid::print(4, 2, ".");
```
<sup><a href='/tests/DocTest_Tests/utilities/GridTests.cpp#L13-L15' title='Snippet source file'>snippet source</a> | <a href='#snippet-grid_simple' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

which produces:

<!-- snippet: GridTests.Print_Simple.approved.txt -->
<a id='snippet-GridTests.Print_Simple.approved.txt'></a>
```txt
....
....
```
<sup><a href='/tests/DocTest_Tests/utilities/approval_tests/GridTests.Print_Simple.approved.txt#L1-L3' title='Snippet source file'>snippet source</a> | <a href='#snippet-GridTests.Print_Simple.approved.txt' title='Start of snippet'>anchor</a></sup>
<!-- endSnippet -->

which you can then build on by gradually implementing your lambda...

---

[Back to User Guide](/doc/README.md#top)
