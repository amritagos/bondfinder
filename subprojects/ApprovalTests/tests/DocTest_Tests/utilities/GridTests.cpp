#include "doctest/doctest.h"
#include "ApprovalTests/Approvals.h"
#include "ApprovalTests/utilities/Grid.h"

TEST_CASE("Print Grid")
{
    auto grid = ApprovalTests::Grid::print(
        6, 8, [](int x, int y, auto& os) { os << ((x % 2 == y % 2) ? "O" : "X"); });
    ApprovalTests::Approvals::verify(grid);
}
TEST_CASE("Print Simple")
{
    // begin-snippet: grid_simple
    auto grid = ApprovalTests::Grid::print(4, 2, ".");
    // end-snippet
    ApprovalTests::Approvals::verify(grid);
}

TEST_CASE("Print Coordinates")
{
    // begin-snippet: grid_lambda
    auto grid = ApprovalTests::Grid::print(
        4, 3, [](int x, int y, auto& os) { os << "(" << x << "," << y << ") "; });
    // end-snippet
    ApprovalTests::Approvals::verify(grid);
}
