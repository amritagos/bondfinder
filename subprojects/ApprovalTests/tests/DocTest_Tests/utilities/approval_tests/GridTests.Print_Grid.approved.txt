OXOXOX
XOXOXO
OXOXOX
XOXOXO
OXOXOX
XOXOXO
OXOXOX
XOXOXO

